### PR TITLE
Configure JaCoCo

### LINKED ISSUE #6048: Integrate test code coverage measurement tool
**Describe the bug**
In order to make tasks like #6047 easier and more efficient, we should integrate code coverage measurement tool so we can spot parts of code that are not covered by any unit test.

**Expected behavior**
Developer is able to execute a command that will result with information about code coverage for specific classes.

### PR DESCRIPTION
**Pull request by wzieba**
Closes: #6048 

Internal communication links: p91TBi-7v0-p2 p91TBi-7Sy-p2

cc: @thehenrybyrd 

### Description
This PR integrates JaCoCo - a tool that will allow us to measure code coverage and make tasks like #6047 easier.

This is more of an experiment for now to see how things work. We do not set any limits for code coverage or expectations. Also JaCoCo won't interfere with any of the existing tools/processes - it's a safe experiment.

#### Why not other tool, e.g. [Kover](https://github.com/Kotlin/kotlinx-kover)?
1. Requires more configuration (e.g. filtering out specific build types/flavours. See [issues](https://github.com/Kotlin/kotlinx-kover/issues?q=is%3Aissue+is%3Aopen+android))
2. Kover is not supported by a tool that can help us measure coverage over time - CodeCov.

#### Why not using code coverage results from Android Studio?
While it's possible, it has some flaws. E.g. we can't get those results without Android Studio/IntelliJ so it's not possible to get results on CI (and then send it to CodeCov). Another thing is that exclusion rules would've been standardised between Android Studio configurations on each team member machine. It adds not worth effort IMO.

### Testing instructions
1. Please run `./gradlew jacocoTestReport`
2. Open html report at `WooCommerce/build/reports/jacoco/wasabiDebug/index.html`
3. Browse it. Make sure that there's no sign of generated code.
4. Assert that code from `cardreader` module is also presented in report (package `com.woocommerce.android.cardreader`)

### Images/gif

image
(...)
image

- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -186,3 +187,29 @@ buildScan {
         uploadInBackground = false
     }
 }
+
+junitJacoco {
+    excludes = [
```
Comment (wzieba) [inline build.gradle]: How I created this list?

1. I took [list](https://github.com/vanniktech/gradle-android-junit-jacoco-plugin/blob/master/src/main/groovy/com/vanniktech/android/junit/jacoco/JunitJacocoExtension.groovy#L24) from the integrated Gradle Plugin
2. I executed code coverage raport
3. I started to remove redundant items like `'**/*AutoValue_*.*'` and running code coverage statistics to see if there's no change
4. Then I browsed code coverage results to identify generated code not included in those exclusion rules (starting from line 210)

**Review by ParaskP7 — COMMENTED** [review]
👋 @wzieba !

I reviewed and test this PR, everything LGTM, great job on adding Jacoco to the project and on making sure to exlclude the necessary files so that there's no sign of generated code to mess up with the final code coverage result! ❤️ 🌟 🚀 

> Why not other tool, e.g. [Kover](https://github.com/Kotlin/kotlinx-kover)?
> - Requires more configuration (e.g. filtering out specific build types/flavours. See [issues](https://github.com/Kotlin/kotlinx-kover/issues?q=is%3Aissue+is%3Aopen+android))
> - Kover is not supported by a tool that can help us measure coverage over time - CodeCov.

Good call. 👍 💯  

-----

I would approve this PR but first I want to pick you brain and understand why you chose to use this [gradle-android-junit-jacoco-plugin](https://github.com/vanniktech/gradle-android-junit-jacoco-plugin) in order to add Jacoco, that is instead of configuring Jacoco directly? 🤔 

I am asking this question because I am usually in favor of not using such contributor plugins unless it is absolutely necessary. Also, the fact that the latest version of Jacoco is [0.8.7 (05 May 2021)](https://github.com/jacoco/jacoco/releases/tag/v0.8.7), while the latest version of this contributor plugin is [0.16.0 (22 Mar 2020)](https://github.com/vanniktech/gradle-android-junit-jacoco-plugin/releases/tag/0.16.0), which is 2 years ago, with a 1 year diff between them, makes me curious whether it would be wiser to use Jacoco directly instead of depending on this custom contributor plugin. Wdyt?

In addition to that, and the [gradle-android-junit-jacoco-plugin](https://github.com/vanniktech/gradle-android-junit-jacoco-plugin) contributor plugin that you decided to use, which I wasn't aware of, I am aware of another such contributor plugin, the [jacoco-android-gradle-plugin](https://github.com/arturdm/jacoco-android-gradle-plugin), with a more up-to-date release, the [0.1.5 (29 Jun 2021)](https://github.com/arturdm/jacoco-android-gradle-plugin/releases/tag/jacoco-android-release-0.1.5). Did you consider this contributor plugin as well when deciding on using a custom contributor plugin?

**Comment by wzieba** [comment]
👋 Thanks for the review!

I'm not in opposition to not use the plugin. Let me try to POC it and we'll see how much code it adds and then decide, WDYT?

Below I'm presenting my arguments to concerns:

> I would approve this PR but first I want to pick you brain and understand why you chose to use this [gradle-android-junit-jacoco-plugin](https://github.com/vanniktech/gradle-android-junit-jacoco-plugin) in order to add Jacoco, that is instead of configuring Jacoco directly? 🤔

The "pure" Jacoco Gradle Plugin doesn't generate tasks for every build variant, while the plugin [does it](https://github.com/vanniktech/gradle-android-junit-jacoco-plugin/blob/master/src/main/groovy/com/vanniktech/android/junit/jacoco/GenerationPlugin.groovy#L143) - I believe it's very convenient that we have all of those tasks out of the box. This and rest of the configuration (like `kotlin/java` directories). We can do this ourselves of course but I thought it's easier and less boiler-plate to use plugin.

> Also, the fact that the latest version of Jacoco is [0.8.7 (05 May 2021)](https://github.com/jacoco/jacoco/releases/tag/v0.8.7), while the latest version of this contributor plugin is [0.16.0 (22 Mar 2020)](https://github.com/vanniktech/gradle-android-junit-jacoco-plugin/releases/tag/0.16.0), which is 2 years ago, with a 1 year diff between them, makes me curious whether it would be wiser to use Jacoco directly instead of depending on this custom contributor plugin.

Good point - thankfully the plugin offers a way to externally set version of Jacoco - [currently](https://github.com/vanniktech/gradle-android-junit-jacoco-plugin/blob/8f2178893544df8337d4292b754b3364bffc985a/src/main/groovy/com/vanniktech/android/junit/jacoco/JunitJacocoExtension.groovy#L12) it's `0.8.2`, we can set it to `0.8.7`

> In addition to that, and the [gradle-android-junit-jacoco-plugin](https://github.com/vanniktech/gradle-android-junit-jacoco-plugin) contributor plugin that you decided to use, which I wasn't aware of, I am aware of another such contributor plugin, the [jacoco-android-gradle-plugin](https://github.com/arturdm/jacoco-android-gradle-plugin), with a more up-to-date release, the [0.1.5 (29 Jun 2021)](https://github.com/arturdm/jacoco-android-gradle-plugin/releases/tag/jacoco-android-release-0.1.5). Did you consider this contributor plugin as well when deciding on using a custom contributor plugin?

Yes I did consider the second plugin you mention! I selected `gradle-android-junit-jacoco-plugin` for following reasons:
- last commits are from ~4 days ago vs. on other plugin it's ~6 months 
- the author of `gradle-android-junit-jacoco-plugin` is a known author of other tools/contributions, usually Gradle related - I have trust in their solution

I don't understand though why the last version is released on 22 Mar 2020 with clear changes comparing to current `master` branch.

**Comment by ParaskP7** [comment]
Thanks for the response @wzieba ! 🙇 

> I'm not in opposition to not use the plugin. Let me try to POC it and we'll see how much code it adds and then decide, WDYT?

Definitely, this sounds like a good plan and will make it easier for us to decide afterwards. 🥇 

> The "pure" Jacoco Gradle Plugin doesn't generate tasks for every build variant, while the plugin [does it](https://github.com/vanniktech/gradle-android-junit-jacoco-plugin/blob/master/src/main/groovy/com/vanniktech/android/junit/jacoco/GenerationPlugin.groovy#L143) - I believe it's very convenient that we have all of those tasks out of the box. This and rest of the configuration (like kotlin/java directories). We can do this ourselves of course but I thought it's easier and less boiler-plate to use plugin.

This is true, but do we really need code coverage for every variant? Maybe choosing to have code coverage only for `wasabiDebug` is enough, at least for now, wdyt? 🤔 

> Good point - thankfully the plugin offers a way to externally set version of Jacoco - [currently](https://github.com/vanniktech/gradle-android-junit-jacoco-plugin/blob/8f2178893544df8337d4292b754b3364bffc985a/src/main/groovy/com/vanniktech/android/junit/jacoco/JunitJacocoExtension.groovy#L12) it's 0.8.2, we can set it to 0.8.7

This is good, an advantage indeed! 🥇 

> Yes I did consider the second plugin you mention! I selected gradle-android-junit-jacoco-plugin for following reasons:
> 
> - last commits are from ~4 days ago vs. on other plugin it's ~6 months
> - the author of gradle-android-junit-jacoco-plugin is a known author of other tools/contributions, usually Gradle related - I have trust in their solution

Great, it is good that you considered other plugins as well! 🥇 

- The last commit on the other plugin is on the [20th of Jan 2022](https://github.com/arturdm/jacoco-android-gradle-plugin/commit/aac810a01451e3df6a9a2a34b6a43e0213384ea8), not ~6 months old, isn't it, what am I missing? 🤔 But anyway, the release version is what we are interested in. No matter what, I understand your point, that this plugin you chose seems to be kept up-to-date, which is a good sign. 👍 
- Although to be honest I wasn't aware of the author, I trust your decision on that. My main and only concern is with this kind of contributor related plugins in general, that is whether they will continue support this plugin going forward. But this begs another questions, do we need that plugin to be that active, maybe not, maybe we will fine with this version indefinitely, but still, I wanted to have us think about that before committing. 👍 

> I don't understand though why the last version is released on 22 Mar 2020 with clear changes comparing to current master branch.

True... 🤔

**Comment by wzieba** [comment]
@ParaskP7 it was a good call with JaCoCo version. Apparently between `0.8.2` and `0.8.7` there's many improvements specifically for Kotlin ([changelog](https://www.jacoco.org/jacoco/trunk/doc/changes.html)) so the report is now much more readable!

WDYT about current solution with plain JaCoCo plugin?

---

While working on it I also tested once Kover again as it's possible to use JaCoCo engine (and this way make it possible to work with CodeCov I believe) but number of issues I had is really overwhelming. It's more of an early alpha to me rather than something we could use. On this branch: https://github.com/woocommerce/woocommerce-android/compare/issue/6048_kover_experiments?expand=1 you can run tasks like `koverMergedHtmlReport` but they won't work (at least didn't for me).

**Comment by ParaskP7** [comment]
👋 @wzieba !

> @ParaskP7 it was a good call with JaCoCo version. Apparently between 0.8.2 and 0.8.7 there's many improvements specifically for Kotlin ([changelog](https://www.jacoco.org/jacoco/trunk/doc/changes.html)) so the report is now much more readable!

🎉 Yeah, I also saw lots of Kotlin googies on the changlog, I am glad that you could diff the report and identify improvements between the two solutions. 🥇 

> WDYT about current solution with plain JaCoCo plugin?

Loving it, thanks for making the extra effort in converting the custom plugin to a pure `Jacoco` solution! 🙇 

I just have 3 suggestions for you:
- Since I am seeing you also want the `libs:cardreader` module to be configured with `Jacoco`, consider adding the `jacoco` plugin there too. Isn't that needed there too? I wonder how is it working for this module already as I am seeing that the code coverage report includes those packages as well... Maybe since `jacoco` is applies on the root level it is not needed on all the module levels, not sure. 🤔 
- Consider moving the `jacoco` configuration to a seperate Gradle file, maybe a `jacoco.gradle`, then add that to a new `config/gradle` folder, and then apply it on the parent and root level `build.gradle` file. I know that we are not currently using this pattern of keeping `build.gradle` files organized and separated like that, but it might be a good start as these root level builds files (and not only) are getting bigger and bigger. Also, considering this is an experiment and might be afterwards removed, instead of removing multiple lines on the root level `build.gradle`, and making sure everything `Jacoco` related is removed, it might be preferred and easier for anyone if just this `apply <file>` line is removed instead, which will make sure everything `Jacoco` related is removed. Wdyt? 🤔 
- Consider updating, or adding a new `Testing Instructions` section for plain `Jacoco`, just for future reference, and in case anyone would want to understand how this, or both solutions, were tested.

> While working on it I also tested once Kover again as it's possible to use JaCoCo engine (and this way make it possible to work with CodeCov I believe) but number of issues I had is really overwhelming. It's more of an early alpha to me rather than something we could use. On this branch: https://github.com/woocommerce/woocommerce-android/compare/issue/6048_kover_experiments?expand=1 you can run tasks like koverMergedHtmlReport but they won't work (at least didn't for me).

Yes, I still feel that we can proceed with `Jacoco` instead of `Kover`, so I totally agree. 👍 No matter, thanks so much for expiremeting with this code coverage tool as well! 🥇

**Comment by wzieba** [comment]
Good suggestions @ParaskP7 !

> I wonder how is it working for this module already as I am seeing that the code coverage report includes those packages as well... Maybe since jacoco is applies on the root level it is not needed on all the module levels, not sure. 🤔

Yes, that's true. As root tasks specifies source for code and compiled classes, there's no need for `subprojects` to know anything about `Jacoco` (which is nice - I believe it's a build script responsibility to handle such tools IMO).

I removed unnecessary application of plugin in a91c8b85772db66fa22efb4c582823d9a5cb000e

> Consider moving the jacoco configuration to a seperate Gradle file, maybe a jacoco.gradle, then add that to a new config/gradle folder

I agree with your arguments. Applied your suggestion in 22f6d61eaecdd16dac89eea25b3eaadf7e16560a.

As a side note: to me, the perfect solution would be to use `buildSrc` declared plugins, which we could write in Kotlin (and unit-test). But that's something for future.

> Consider updating, or adding a new Testing Instructions section for plain Jacoco

Done ✅.

**Comment by ParaskP7** [comment]
👋 @wzieba !

Thank you so much for applying the suggestions, you 🤘 , all 🟢 from my side!

> Yes, that's true. As root tasks specifies source for code and compiled classes, there's no need for subprojects to know anything about Jacoco (which is nice - I believe it's a build script responsibility to handle such tools IMO).

👍 

> I removed unnecessary application of plugin in https://github.com/woocommerce/woocommerce-android/commit/a91c8b85772db66fa22efb4c582823d9a5cb000e

❤️ 

> I agree with your arguments. Applied your suggestion in https://github.com/woocommerce/woocommerce-android/commit/22f6d61eaecdd16dac89eea25b3eaadf7e16560a.

❤️ 

> As a side note: to me, the perfect solution would be to use buildSrc declared plugins, which we could write in Kotlin (and unit-test). But that's something for future.

💯 

> Done ✅.

🙇

**Review by ParaskP7 — APPROVED** [review]
LGTM! 💯

**Pull request merged**